### ISSUE TITLE
[Bug]: Sometimes lemmy returns the wrong rich link preview

### ISSUE THREAD

**Issue by db0** [LemmyNet/lemmy-ui#2808]
### Requirements

- [X] Is this a bug report? For questions or discussions use https://lemmy.ml/c/lemmy_support
- [X] Did you check to see if this issue already exists?
- [X] Is this only a single bug? Do not put multiple bugs in one issue.
- [X] Do you agree to follow the rules in our [Code of Conduct](https://join-lemmy.org/docs/code_of_conduct.html)?
- [X] Is this a backend issue? Use the [lemmy-ui](https://github.com/LemmyNet/lemmy-ui) repo for UI / frontend issues.

### Summary

This is something I noticed the last week. Sometimes when linking to a lemmy post, the wrong preview image and blurb is being returned. Initially I thought it was just a fluke when I [linked to this post](https://lemmy.dbzer0.com/post/31168360) and got a different result in the preview, but today it happened a second time from mastodon, with pretty catastrophic results. I posted one link to mastodon, and the preview was a completely different and **very** NSFW preview instead. The only way to "fix" it was to edit my post and add a random query at the end.

I asked on matrix and it seems I'm not the only one affected https://mstdn.social/@DJxSpeedy/113443024738078089

I checked manually on my post and the preview meta tags are all, currently correc.

Not sure if this is a UI issue or a backend issue. Feel free to move.

### Steps to Reproduce

1. Post a lemmy link somewhere which fetches link previews.
2. Randomly get the wrong one

### Technical Details

I'm on lemmy 0.19.6 but this seemns to have also affected previous versions

### Version

0.19.6

### Lemmy Instance URL

lemmy.dbzer0.com

**Comment by dessalines** [comment]
I don't think this can be a lemmy-ui issue, because the only metadata fetching it does, is to fetch a page title when creating a post. The thumbnails, embed title, and embed description are all fetched in the back-end, from the URL.

I've seen these cases only when you create a post with the wrong url, then correct it afterwards.

The example above is pretty strange, and must have been done not using the cross-post button, since instead of using the https://join-lemmy.org/news/2024-11-08_-_Lemmy_Release_v0.19.6 url, it used a different lemmy link. My guess is that the original URL was the ambulance one, then you tried to correct it but used a different internal lemmy link instead of the join-lemmy.org link.

Its tough to follow but I'd need a reproducible example.

**Comment by db0** [comment]
The ones I mentioned are certainly not the wrong url and then edited, as they are things I've never even seen myself.

I'll post here when it happens to me.

**Comment by VictorFloret** [comment]
Seems like the problem is that the message created by the lemmy's "share" button contais the post's body in it, and the link to the post is added below the body's text. If the body contains any other URL, apps like WhatsApp and Telegram will show a preview to that URL, and not the post's URL.

![Screenshot_20241218_165853_Firefox](https://github.com/user-attachments/assets/fc90f6de-700f-454c-9b7c-26ee3ff092aa)
![Screenshot_20241218_165917_Telegram](https://github.com/user-attachments/assets/db7d32fb-fdf9-42b1-9050-c797937f67ac)

>I posted one link to mastodon, and the preview was a completely different and very NSFW preview instead. 

This probably happens because the URL in the body can get truncated, resulting in a different URL.

**Comment by VictorFloret** [comment]
> Seems like the problem is that the message created by the lemmy's "share" button contais the post's body in it, and the link to the post is added below the body's text. If the body contains any other URL, apps like WhatsApp and Telegram will show a preview to that URL, and not the post's URL.

A solution to this would be adding the post's body thest AFTER the post's URL.

**Comment by Nothing4You** [comment]
I just saw this happening on a feddit.uk (Lemmy 0.19.7) post that I was linking elsewhere:

https://feddit.uk/post/21568696

<img width="1526" alt="image" src="https://github.com/user-attachments/assets/621d4573-3972-44e7-a0ff-a837aec6cadf" />

This post actually got multiple different rich previews within a few seconds of each other and it is not edited.

The first place where I linked this was showing a text preview for what appears to have been 10 month old post https://feddit.uk/post/8913602, at least the title and description match. There was also an image preview that I don't see on that post.

When I noticed the wrong preview I opened the post in my browser, looked at the page source, and I saw metadata for an entirely different post (screenshot above). That time it returned metadata of 2 month old post https://feddit.uk/post/19003214. This is also confirmed by the html containing `<link data-inferno-helmet="true" rel="canonical" href="https
://feddit.uk/post/19003214">`.

When I used curl to retrieve the page again separately or reloaded again in my browser it returned correct metadata.

The HTML did not have any relevant changes other than title, meta tags and the canonical link.

**Comment by ValueSubtracted** [comment]
I think this sounds similar to something I sometimes see when viewing Lemmy posts from Mastodon accounts. As an added wrinkle, it doesn't appear that way on **all** Mastodon accounts.

Sometimes, it displays a false preview on Mastodon Server A, but not Mastodon Server B.

Another post might look correct on Server A, and wrong on Server B.

And often, posts look completely fine on both.

Here's a screenshot of a post that is displaying correctly on one Mastodon server:

<img width="579" height="486" alt="Image" src="https://github.com/user-attachments/assets/ae0ef12e-2666-458a-89db-fcca3719942d" />

And here's that same post seen on a different Mastodon server, with a false link preview attached:

<img width="367" height="716" alt="Image" src="https://github.com/user-attachments/assets/f379fed5-257b-4e57-99a8-625f99766ded" />

If you click on the "false" link (in this example, the Gundam thing), you will be redirected to the correct post (in this case, the episode discussion).

I unfortunately don't know how to repro this consistently.

**Comment by dessalines** [comment]
It could be a caching issue, solved by #3248

**Comment by Nutomic** [comment]
Has anyone noticed this bug with Lemmy 0.19.13 or later? That is when the fix above was released. If the problem is gone then we can consider this fixed.

**Comment by db0** [comment]
I haven't seen it myself but I haven't cross-posted elsewhere a lot lately either.

**Comment by ValueSubtracted** [comment]
Unfortunately, yes, though my instance is not involved.

Lemmy post [here](https://sh.itjust.works/post/56621224) - it was posted to lemmy.ca/c/canada

What it looks like on my Mastodon server:

<img width="367" height="743" alt="Image" src="https://github.com/user-attachments/assets/5ca97213-157b-4b58-aa7d-88388560a03d" />

It seems to be pulling the preview from either [here](https://sh.itjust.works/post/55047956) or [here](https://sh.itjust.works/post/55141916).

Both lemmy.ca and sh.itjust.works are running Lemmy 0.19.16, and the Mastodon server is running  v4.6.0-alpha.5+glitch

**Comment by Nutomic** [comment]
Damn, its extremely difficult to solve this without any way to reproduce it locally. Interestingly, in that screenshot the preview description is correct, only image and title are wrong.

Most likely there is some sort of race condition, or it uses stale data from a previous request. But the corresponding code in `post.tsx` seems to initialize the data correctly, and checks for request success before rendering anything.

**Comment by Nutomic** [comment]
Another possibility is that the problem is caused by `inferno-helmet` which manages this metadata. The library is written in plain JS without any lints or type checks, and is unmaintained since 8 years. It also doesnt support the latest Inferno version and has very few users, so theres a high chance that it has some bugs. We should really remove it and put the code directly into lemmy-ui (https://github.com/LemmyNet/lemmy-ui/issues/3881).

Edit: inferno-helmet uses [inferno-side-effect](https://github.com/Schiavi/inferno-side-effect) which stores component prop data in static variables. I suspect that it's responsible for this bug.
